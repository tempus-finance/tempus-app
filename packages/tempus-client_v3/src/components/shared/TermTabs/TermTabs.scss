.tc__term-tabs {
  height: 76px;
  display: flex;
}

.tc__term-tabs__term {
  margin-right: 8px;
  display: flex;
  flex: 1;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  background-color: var(--termTabsBackground);
  border: 1px solid var(--termTabsBorder);
  box-sizing: border-box;
  border-radius: 4px;

  &:last-child {
    margin-right: 0;
  }
}

.tc__term-tabs__term-selected {
  background-color: var(--termTabsSelectedBackground);
  border-color: var(--termTabsSelectedBorder);
}
