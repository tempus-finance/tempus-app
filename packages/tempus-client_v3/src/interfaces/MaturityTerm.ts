import { Numberish } from 'tempus-core-services';

export default interface MaturityTerm {
  apr: Numberish;
  date: Date;
}
