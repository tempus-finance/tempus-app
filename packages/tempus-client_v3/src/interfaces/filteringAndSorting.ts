export type PoolType = 'fixed' | 'boosted' | 'all';
export type ViewType = 'grid' | 'list';
export type FilterType = 'active' | 'matured' | 'inactive';
export type SortType = 'a-z' | 'maturity' | 'tvl' | 'apr' | 'balance';
export type SortOrder = 'asc' | 'desc';
