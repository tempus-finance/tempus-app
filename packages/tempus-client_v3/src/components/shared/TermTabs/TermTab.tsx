import { FC, memo, useCallback } from 'react';
import { DecimalUtils } from 'tempus-core-services';
import ButtonWrapper from '../ButtonWrapper';
import FormattedDate from '../FormattedDate';
import Typography from '../Typography';
import MaturityTerm from './MaturityTerm';

interface TermTabProps {
  term: MaturityTerm;
  selected: boolean;
  onClick: (selected: MaturityTerm) => void;
}

const TermTab: FC<TermTabProps> = props => {
  const { term, selected, onClick } = props;
  const handleClick = useCallback(() => onClick(term), [term, onClick]);

  return (
    <ButtonWrapper
      className={`tc__term-tabs__term ${selected ? 'tc__term-tabs__term-selected' : ''}`}
      onClick={handleClick}
    >
      <Typography variant="body-secondary" weight="medium" color="text-secondary">
        APR &amp; Term
      </Typography>
      <Typography variant="body-primary" type="mono" weight="bold">
        {DecimalUtils.formatPercentage(term.apr, 1)}
      </Typography>
      <FormattedDate date={term.date} textColor="text-secondary" size="small" separatorContrast="low" />
    </ButtonWrapper>
  );
};

export default memo(TermTab);
