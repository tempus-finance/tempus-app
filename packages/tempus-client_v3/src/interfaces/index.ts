import MaturityTerm from './MaturityTerm';

export type { FilterType, PoolType, SortOrder, SortType, ViewType } from './filteringAndSorting';
export type { TokenMetadataProp } from './TokenMetadata';
export type { MaturityTerm };
