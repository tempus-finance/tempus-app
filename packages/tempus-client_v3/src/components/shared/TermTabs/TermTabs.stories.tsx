import { ComponentMeta } from '@storybook/react';
import { FC } from 'react';
import TermTabs from './TermTabs';

export default {
  title: 'TermTabs',
  component: TermTabs,
  argTypes: {},
} as ComponentMeta<typeof TermTabs>;

export const TwoTermDates: FC = () => <TermTabs dates={[new Date(2022, 7, 1), new Date(2022, 9, 1)]} />;

export const ThreeTermDates: FC = () => (
  <TermTabs dates={[new Date(2022, 7, 1), new Date(2022, 9, 1), new Date(2022, 11, 1)]} />
);
