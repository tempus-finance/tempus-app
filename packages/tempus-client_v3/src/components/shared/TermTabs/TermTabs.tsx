import { FC, useCallback, useEffect, useState } from 'react';
import { MaturityTerm } from '../../../interfaces';
import TermTab from './TermTab';
import './TermTabs.scss';

export interface TermTabsProps {
  terms: MaturityTerm[];
  disabled?: boolean;
  onChange?: (selected: MaturityTerm) => void;
}

const TermTabs: FC<TermTabsProps> = props => {
  const { terms, disabled, onChange } = props;
  const [selectedTerm, setSelectedTerm] = useState<MaturityTerm>(terms[0]);
  const onClick = useCallback(
    (value: MaturityTerm) => {
      if (terms.length > 1) {
        setSelectedTerm(value);
        onChange?.(value);
      }
    },
    [terms, onChange],
  );

  useEffect(() => {
    if (terms.length === 1) {
      setSelectedTerm(terms[0]);
      onChange?.(terms[0]);
    }
  }, [terms, onChange]);

  return (
    <div className="tc__term-tabs">
      {terms.map((term, index) => (
        <TermTab
          term={term}
          selected={selectedTerm === term}
          disabled={disabled}
          onClick={onClick}
          key={`term-tab-${index}-${term.date.getTime()}`}
        />
      ))}
    </div>
  );
};

export default TermTabs;
