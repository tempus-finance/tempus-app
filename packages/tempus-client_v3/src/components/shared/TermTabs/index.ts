import MaturityTerm from './MaturityTerm';
import TermTabs from './TermTabs';

export default TermTabs;

export type { MaturityTerm };
