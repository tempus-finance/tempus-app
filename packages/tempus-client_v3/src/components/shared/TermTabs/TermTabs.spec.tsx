import { fireEvent, render } from '@testing-library/react';
import I18nProvider from '../../../i18n/I18nProvider';
import { MaturityTerm } from '../../../interfaces';
import TermTabs, { TermTabsProps } from './TermTabs';

const mockOnChange = jest.fn<void, [MaturityTerm]>();

const subject = (props: TermTabsProps) =>
  render(
    <I18nProvider>
      <TermTabs {...props} />
    </I18nProvider>,
  );

describe('TermTabs', () => {
  it('renders a two-button term tabs', () => {
    const { baseElement, getAllByRole } = subject({
      terms: [
        {
          apr: 0.042,
          date: new Date(2022, 0, 1),
        },
        {
          apr: 0.1,
          date: new Date(2022, 5, 1),
        },
      ],
    });

    const buttons = getAllByRole('button');
    buttons.forEach(button => expect(button).not.toBeNull());

    expect(baseElement).toMatchSnapshot();
  });

  it('calls `onChange` when a term tab is clicked', () => {
    const terms = [
      {
        apr: 0.042,
        date: new Date(2022, 0, 1),
      },
      {
        apr: 0.1,
        date: new Date(2022, 5, 1),
      },
      {
        apr: 0.125,
        date: new Date(2022, 10, 1),
      },
    ];
    const { getAllByRole } = subject({
      terms,
      onChange: mockOnChange,
    });

    const buttons = getAllByRole('button');
    fireEvent.click(buttons[1]);

    expect(mockOnChange).toHaveBeenCalledTimes(1);
    expect(mockOnChange).toHaveBeenCalledWith(terms[1]);
  });
});
