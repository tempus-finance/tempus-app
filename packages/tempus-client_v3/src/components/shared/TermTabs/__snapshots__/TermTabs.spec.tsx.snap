// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TermTabs renders a two-button term tabs 1`] = `
<body>
  <div>
    <div
      class="tc__term-tabs"
    >
      <button
        class="tc__btn tc__term-tabs__term tc__term-tabs__term-selected"
        data-selected="false"
        type="button"
      >
        <div
          style="font-style: normal; font-size: 12px; line-height: 16px; color: rgb(34, 34, 34); font-weight: 500; font-family: 'DM Sans', sans-serif; opacity: 1;"
        >
          Term
        </div>
        <div
          class="tc__term-tabs__term-date"
          style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 400; font-family: 'Azeret Mono', monospace; opacity: 1;"
        >
          <span
            class="tc__formatted-date"
          >
            <div
              style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 500; font-family: 'Azeret Mono', monospace; opacity: 1;"
            >
              Jan
            </div>
            <span
              class="tc__formatted-date__separator tc__formatted-date__separator-high-contrast"
            />
            <div
              style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 500; font-family: 'Azeret Mono', monospace; opacity: 1;"
            >
              01
            </div>
            <span
              class="tc__formatted-date__separator tc__formatted-date__separator-high-contrast"
            />
            <div
              style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 500; font-family: 'Azeret Mono', monospace; opacity: 1;"
            >
              22
            </div>
          </span>
        </div>
      </button>
      <button
        class="tc__btn tc__term-tabs__term "
        data-selected="false"
        type="button"
      >
        <div
          style="font-style: normal; font-size: 12px; line-height: 16px; color: rgb(34, 34, 34); font-weight: 500; font-family: 'DM Sans', sans-serif; opacity: 1;"
        >
          Term
        </div>
        <div
          class="tc__term-tabs__term-date"
          style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 400; font-family: 'Azeret Mono', monospace; opacity: 1;"
        >
          <span
            class="tc__formatted-date"
          >
            <div
              style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 500; font-family: 'Azeret Mono', monospace; opacity: 1;"
            >
              Jun
            </div>
            <span
              class="tc__formatted-date__separator tc__formatted-date__separator-high-contrast"
            />
            <div
              style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 500; font-family: 'Azeret Mono', monospace; opacity: 1;"
            >
              01
            </div>
            <span
              class="tc__formatted-date__separator tc__formatted-date__separator-high-contrast"
            />
            <div
              style="font-style: normal; font-size: 16px; line-height: 24px; color: rgb(34, 34, 34); font-weight: 500; font-family: 'Azeret Mono', monospace; opacity: 1;"
            >
              22
            </div>
          </span>
        </div>
      </button>
    </div>
  </div>
</body>
`;
